pkg/kube/krt/mergejoin.go

```diff
@@ -0,0 +1,315 @@
+			continue
+		}
+		res = append(res, v)
+	}
+	return res
+}
+
+func (c joinCollectionIndex[T]) delete(o T, oKey Key[T]) {
+	oldIndexKeys := c.extract(o)
+	for _, oldIndexKey := range oldIndexKeys {
+		sets.DeleteCleanupLast(c.index, oldIndexKey, oKey)
+	}
+}
+
+func (c joinCollectionIndex[T]) update(ev Event[T], oKey Key[T]) {
+	if ev.Old != nil {
+		c.delete(*ev.Old, oKey)
+	}
+	if ev.New != nil {
+		newIndexKeys := c.extract(*ev.New)
+		for _, newIndexKey := range newIndexKeys {
+			sets.InsertOrNew(c.index, newIndexKey, oKey)
+		}
+	}
+}
+
+func (j *mergejoin[T]) Metadata() Metadata {
+	return j.metadata
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (j *mergejoin[T]) augment(a any) any {
+	// not supported in this collection type
+	return a
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (j *mergejoin[T]) name() string { return j.collectionName }
+
+// nolint: unused // (not true, its to implement an interface)
+func (j *mergejoin[T]) uid() collectionUID { return j.id }
+
+func (j *mergejoin[T]) Synced() Syncer {
+	return channelSyncer{
+		name:   j.collectionName,
+		synced: j.synced,
+	}
+}
+
+// nolint: unused // (not true)
+func (j *mergejoin[T]) index(name string, extract func(o T) []string) indexer[T] {
+	j.mu.Lock()
+	defer j.mu.Unlock()
+	if idx, ok := j.indexes[name]; ok {
+		return idx
+	}
+
+	idx := joinCollectionIndex[T]{
+		extract: extract,
+		index:   make(map[string]sets.Set[Key[T]]),
+		parent:  j,
+	}
+	for k, v := range j.outputs {
+		idx.update(Event[T]{
+			Old:   nil,
+			New:   &v,
+			Event: controllers.EventAdd,
+		}, k)
+	}
+	j.indexes[name] = idx
+	return idx
+}
+
+func (j *mergejoin[T]) HasSynced() bool {
+	return j.syncer.HasSynced()
+}
+
+func (j *mergejoin[T]) WaitUntilSynced(s <-chan struct{}) bool {
+	return j.syncer.WaitUntilSynced(s)
+}
+
+func (j *mergejoin[T]) List() []T {
+	j.mu.RLock()
+	defer j.mu.RUnlock()
+
+	return maps.Values(j.outputs)
+}
+
+func (j *mergejoin[T]) GetKey(k string) *T {
+	j.mu.RLock()
+	defer j.mu.RUnlock()
+
+	rf, f := j.outputs[Key[T](k)]
+	if f {
+		return &rf
+	}
+	return nil
+}
+
+func (j *mergejoin[T]) onSubCollectionEventHandler(o []Event[T]) {
+	var events []Event[T]
+	j.mu.Lock()
+	defer j.mu.Unlock()
+	// Between the events being enqueued and now, the input may have changed. Update with latest info.
+
+	items := j.refreshEventsLocked(o)
+
+	for _, ev := range items {
+		obj := ev.Latest()
+		objKey := getTypedKey(obj)
+		if ev.Event == controllers.EventDelete {
+			oldRes, f := j.outputs[objKey]
+			if !f {
+				j.log.WithLabels("iKey", objKey).Errorf("invalid event, deletion of non-existent object")
+				continue
+			}
+			e := Event[T]{
+				Event: controllers.EventDelete,
+				Old:   &oldRes,
+			}
+			events = append(events, e)
+			delete(j.outputs, objKey)
+			for _, index := range j.indexes {
+				index.delete(oldRes, objKey)
+			}
+			if j.log.DebugEnabled() {
```

Comment:
Why do we need to check if debug is enabled? Shouldn't the logging library do that check for us?

Reply:
WithLabels is not guarded by log level, and this could be called super often.